### PR TITLE
Add empty check for types in ORC file

### PR DESCRIPTION
**Pull request by wenleix**

### CODE REVIEW

**Comment by wenleix** [comment]
Today it returns `HIVE_CANNOT_OPEN_SPLIT` with the following stacktrace, which is confusing:

```
com.facebook.presto.spi.PrestoException: Error opening Hive split ... : 0
	at com.facebook.presto.hive.orc.OrcPageSourceFactory.createOrcPageSource(OrcPageSourceFactory.java:237)
	at com.facebook.presto.hive.orc.DwrfPageSourceFactory.createPageSource(DwrfPageSourceFactory.java:83)
	at com.facebook.presto.hive.HivePageSourceProvider.createHivePageSource(HivePageSourceProvider.java:157)
	at com.facebook.presto.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:93)
	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:44)
	at com.facebook.presto.split.PageSourceManager.createPageSource(PageSourceManager.java:56)
	at com.facebook.presto.operator.ScanFilterAndProjectOperator.getOutput(ScanFilterAndProjectOperator.java:216)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:379)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:283)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:675)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:276)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1065)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:483)
	at com.facebook.presto.$gen.Presto_0_213_3_g7d8c41f____20181128_195004_1.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:60)
	at com.facebook.presto.orc.OrcRecordReader.<init>(OrcRecordReader.java:163)
	at com.facebook.presto.orc.OrcReader.createRecordReader(OrcReader.java:255)
	at com.facebook.presto.hive.orc.OrcPageSourceFactory.createOrcPageSource(OrcPageSourceFactory.java:207)
	... 17 more
```

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java [hunk 1]
```diff
@@ -184,6 +184,9 @@ public OrcReader(OrcDataSource orcDataSource, OrcEncoding orcEncoding, DataSize
         try (InputStream footerInputStream = new OrcInputStream(orcDataSource.getId(), footerSlice.getInput(), decompressor, newSimpleAggregatedMemoryContext(), footerSize)) {
             this.footer = metadataReader.readFooter(hiveWriterVersion, footerInputStream);
         }
+        if (footer.getTypes().size() == 0) {
```
Comment (sopel39) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java]: Are such files illegal by ORC standard?

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java [hunk 2]
```diff
@@ -184,6 +184,9 @@ public OrcReader(OrcDataSource orcDataSource, OrcEncoding orcEncoding, DataSize
...
+            throw new OrcCorruptionException(orcDataSource.getId(), "Types in footer cannot be empty");
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java]: This should probably say "File has no columns"

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java (hunk 1 above)
Comment (dain) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java]: The spec doesn't say explicitly say the file must contain types, but it is pretty clear that types are expected.  I suggest we make this change, and if someone comes with a real use case later we can change this code allow this.

Comment (wenleix) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java]: I agree with @dain . My understanding about spec is even when there is no column, the root of the type tree still exists (and is a struct of nothing).

https://orc.apache.org/specification/ORCv0/, section `Type Information`: 

> The type tree is flattened in to a list via a pre-order traversal where each type is assigned the next id. Clearly the root of the type tree is always type id 0. Compound types have a field named subtypes that contains the list of their children’s type ids.

**Pull request merged**